skyd/skyd.go

```diff
@@ -1,16 +1,137 @@
 package skyd
 
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+var (
+	// skylinksCache is a local cache of the list of skylinks pinned by the
+	// local skyd
+	skylinksCache pinnedSkylinksCache
+	// skylinksCacheDuration defines the duration of skylinksCache
+	skylinksCacheDuration = time.Hour
+)
+
 type (
 	// Client allows us to call the local skyd instance.
 	Client struct {
+		staticClient *client.Client
+	}
+	// ClientInterface describes the interface exposed by Client, so we can
+	// mock it for testing.
+	ClientInterface interface {
+		Pin(skylink string) error
+		PinnedSkylinks() (skylinks []string, err error)
+		Unpin(skylink string) error
+	}
+
+	// renterDirCache is a simple cache of the renter's directory information,
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   []string
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) *Client {
+	opts := client.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &Client{
+		staticClient: client.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *Client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	sp, err := skymodules.NewSiaPath(skylink)
+	if err != nil {
+		return errors.AddContext(err, "failed to create siapath")
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: sp,
+	}
+	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
+}
+
+// PinnedSkylinks returns the list of skylinks pinned by the local skyd.
+func (c *Client) PinnedSkylinks() (skylinks []string, err error) {
+	skylinksCache.mu.Lock()
+	defer skylinksCache.mu.Unlock()
+	// Check whether the cache is still valid and return it if so.
+	if skylinksCache.Expiration.After(time.Now().UTC()) {
```

Comment:
`nit` probably better to expose a method on the cache called `IsExpired`

```go
func (slc *skylinksCache) IsExpired() bool {
    slc.mu.Lock()
    defer slc.mu.Unlock()
    return slc.Expiration.Before(time.Now().UTC()
}
```

actually probably you'd want something like

```go
var errCacheExpired = errros.New("cache expired")
func (slc *skylinksCache) Skylinks() ([]string, err) {
    slc.mu.Lock()
    defer slc.mu.Unlock()
    if slc.Expiration.Before(time.Now().UTC() {
        return nil, errCacheExpired
    }
    return slc.skylinks, nil
}

...

func main() {
    skylinks, err := cache.Skylinks()
    if err != nil && errors.Contains(err, errCacheExpired) {
        // rebuild cache and reload skylinks
        cache.Rebuild()
        skylinks, err = cache.Skylinks()
    }
    if err != nil {
        // this is bad
        return err
    }
    // use skylinks
}
```

I would really try to limit the amount of time you hold the cache lock.
Oh and  `Rebuild` needs to flip a flag saying it's rebuilding already, so concurrent processes don't start all rebuilding the cache, you'd want the first to rebuild it and all the others to block on that.

Reply:
This goes hand in hand with https://github.com/SkynetLabs/pinner/pull/7#discussion_r865697163. It seems like something that's easier to discuss on Discord/voice.

Reply:
wondering whether rebuilding the cache should be a background process, we have about 1-2M files on our largest nodes I think? and we split it up with the `aa/bb/cc/` thing on disk, so how many directories are we walking here and how long does that take, not trying to be annoying but it might seriously take a really long time

Reply:
I'm also wondering about the amount of stress this puts on a skyd node, probably not to be underestimated. This might bring a node to a grind. Maybe we should consider reading the actual disk instead of querying skyd as well? This is container <-> container asking about filesystem info essentially right?